presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java

```diff
@@ -0,0 +1,412 @@
+                        assignments);
+            }
+            else {
+                return aggregationNode.get();
+            }
+        }
+
+        private Optional<AggregationNode> createAggregationNode(
+                AggregationNode scalarAggregation,
+                JoinNode leftOuterJoin,
+                Symbol nonNullableAggregationSourceSymbol)
+        {
+            ImmutableMap.Builder<Symbol, FunctionCall> aggregations = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Signature> functions = ImmutableMap.builder();
+            FunctionRegistry functionRegistry = metadata.getFunctionRegistry();
+            for (Map.Entry<Symbol, FunctionCall> entry : scalarAggregation.getAggregations().entrySet()) {
+                FunctionCall call = entry.getValue();
+                QualifiedName count = QualifiedName.of("count");
+                Symbol symbol = entry.getKey();
+                if (call.getName().equals(count)) {
+                    aggregations.put(symbol, new FunctionCall(
+                            count,
+                            ImmutableList.of(nonNullableAggregationSourceSymbol.toSymbolReference())));
+                    List<TypeSignature> scalarAggregationSourceTypeSignatures = ImmutableList.of(
+                            symbolAllocator.getTypes().get(nonNullableAggregationSourceSymbol).getTypeSignature());
+                    functions.put(symbol, functionRegistry.resolveFunction(
+                            count,
+                            scalarAggregationSourceTypeSignatures,
+                            false));
+                }
+                else {
+                    aggregations.put(symbol, entry.getValue());
+                    functions.put(symbol, scalarAggregation.getFunctions().get(symbol));
+                }
+            }
+
+            List<Symbol> groupBySymbols = leftOuterJoin.getLeft().getOutputSymbols();
+            return Optional.of(new AggregationNode(
+                    idAllocator.getNextId(),
+                    leftOuterJoin,
+                    groupBySymbols,
+                    aggregations.build(),
+                    functions.build(),
+                    scalarAggregation.getMasks(),
+                    ImmutableList.of(groupBySymbols),
+                    scalarAggregation.getStep(),
+                    scalarAggregation.getSampleWeight(),
+                    scalarAggregation.getConfidence(),
+                    scalarAggregation.getHashSymbol()));
+        }
+    }
+
+    private static class ScalarAggregationSourceRewriter
+    {
+        private final PlanNodeIdAllocator idAllocator;
+
+        private ScalarAggregationSourceRewriter(PlanNodeIdAllocator idAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+        }
+
+        public Optional<Result> rewrite(AggregationNode scalarAggregation, List<Symbol> correlation)
```

Comment:
This method could be made a bit more readable by untangling some of the things it's doing. Something like:

```java
public Optional<Result> rewrite(AggregationNode aggregation, List<Symbol> boundVariables)
{
    PlanNode child = aggregation.getSource();

    Optional<FilterNode> filter = findFirstPlanNodeMatching(
            child,
            FilterNode.class::isInstance,
            Predicates.<PlanNode>isInstance(ProjectNode.class).or(LimitNode.class::isInstance));

    if (!filter.isPresent()) {
        return Optional.of(new Result(ImmutableList.of(), child));
    }

    Expression predicate = filter.get().getPredicate();

    if (!isSupportedPredicate(predicate)) {
        return Optional.empty();
    }

    if (!DependencyExtractor.extractAll(predicate).containsAll(boundVariables)) {
        return Optional.empty();
    }

    ImmutableList.Builder<Expression> joinExpressions = ImmutableList.builder();
    ImmutableList.Builder<Expression> newPredicate = ImmutableList.builder();

    for (Expression conjunct : extractConjuncts(predicate)) {
        List<Symbol> symbols = DependencyExtractor.extractAll(conjunct);

        if (boundVariables.stream().anyMatch(symbols::contains)) {
            joinExpressions.add(conjunct);
        }
        else {
            newPredicate.add(conjunct);
        }
    }
    ...
}
```